### ISSUE TITLE
Kind error in simple inheritance

### ISSUE THREAD

**Issue by szeiger** [lampepfl/dotty#3467]
We're running into an error compiling collection-strawman on Dotty 0.4.0-RC1. Here's the branch with the reproduction: https://github.com/scala/collection-strawman/compare/wip/dotty-kind-error

Steps to reproduce:

```
# Cleaning in sbt is bugged in our build; do it manually to make sure you rebuild everything
$ rm -r collections/.jvm/target/

$ sbt ++0.4.0-RC1 collectionsJVM/test:compile
[...]
[info] Compiling 86 Scala sources and 5 Java sources to /Users/szeiger/code/scala/collection-strawman/collections/.jvm/target/scala-0.4/classes...
[...]
[warn] 12 warnings found
[info] Compiling 1 Scala source to /Users/szeiger/code/scala/collection-strawman/collections/.jvm/target/scala-0.4/test-classes...
[error]
[error] Type argument strawman.collection.immutable.IndexedSeq has not the same kind as its bound  <: [_$63] => scala.Any
[error] one error found
[error] (collectionsJVM/test:compileIncremental) Compilation failed
```

The error message does not give any position. It is caused by refining the type of `ArrayOps.withFilter`:

```scala
  override def withFilter(p: A => Boolean): ArrayOps.WithFilter[A] = new ArrayOps.WithFilter[A](p, this)
```

Without the type annotation here (or by explicitly annotating it with the original type `collection.WithFilter[A, immutable.IndexedSeq]`) the test class compiles. I have been unable to come up with any workaround that would allow us to refine the type (as required to get some extra methods).

The `IndexedSeq` type argument mentioned in the error message comes from the definition of `ArrayOps.WithFilter`:

```
class WithFilter[A](p: A => Boolean, ao: ArrayOps[A]) extends collection.WithFilter[A, immutable.IndexedSeq]
```

This error is a blocker for https://github.com/scala/collection-strawman/pull/278

**Comment by allanrenucci** [comment]
I was able to reproduce:

In `collection.scala`:
```scala
package collection

abstract class WithFilter[+A, +CC[_]]

trait IndexedSeq[+A] extends Any with IndexedSeqOps[A, IndexedSeq, IndexedSeq[A]]

trait IndexedSeqOps[+A, +CC[X] <: IndexedSeq[X], +C] extends Any {
  def withFilter(p: A => Boolean): WithFilter[A, CC] = ???
}

package immutable {
  trait IndexedSeq[+A] extends collection.IndexedSeq[A] with collection.IndexedSeqOps[A, IndexedSeq, IndexedSeq[A]]
}

object ArrayOps {
  abstract class WithFilter[A] extends collection.WithFilter[A, immutable.IndexedSeq]
}

class ArrayOps[A](val xs: Array[A]) extends AnyVal with IndexedSeqOps[A, immutable.IndexedSeq, Array[A]] {
  override def withFilter(p: A => Boolean): ArrayOps.WithFilter[A] = ???
}
```

In `Test.scala`:
```scala
package collection

class Test {
  def test(xs: Array[Int]): Unit = {
    new ArrayOps(xs)
  }
}
```
```
dotty(master|…) ❯ ./bin/dotc collection.scala; ./bin/dotc Test.scala

Type argument collection.immutable.IndexedSeq has not the same kind as its bound  <: [_$1] => Any

one error found
```

**Comment by allanrenucci** [comment]
@szeiger I tested the fix on `wip/dotty-kind-error` and confirms that the tests now compile. The fix will be in Dotty 0.5.0-RC1 which will be released on Wednesday
